mods/default/functions.lua

```diff
@@ -353,32 +353,71 @@ minetest.register_abm({
 
 
 --
--- Grass growing on well-lit dirt
+-- convert dirt to something that fits the environment
 --
 
 minetest.register_abm({
 	nodenames = {"default:dirt"},
-	neighbors = {"air"},
+	neighbors = {
+		"default:dirt_with_grass",
+		"default:dirt_with_dry_grass",
+		"default:dirt_with_snow",
+		"default:grass_1",
+		"default:grass_2",
+		"default:grass_3",
+		"default:grass_4",
+		"default:grass_5",
+		"default:dry_grass_1",
+		"default:dry_grass_2",
+		"default:dry_grass_3",
+		"default:dry_grass_4",
+		"default:dry_grass_5",
+		"default:snow",
+	},
 	interval = 6,
 	chance = 67,
 	catch_up = false,
 	action = function(pos, node)
+		-- most likely case, half the time it's too dark for this
 		local above = {x = pos.x, y = pos.y + 1, z = pos.z}
-		local name = minetest.get_node(above).name
-		local nodedef = minetest.registered_nodes[name]
-		if nodedef and (nodedef.sunlight_propagates or nodedef.paramtype == "light") and
-				nodedef.liquidtype == "none" and
-				(minetest.get_node_light(above) or 0) >= 13 then
-			if name == "default:snow" or name == "default:snowblock" then
-				minetest.set_node(pos, {name = "default:dirt_with_snow"})
-			else
-				minetest.set_node(pos, {name = "default:dirt_with_grass"})
+		if (minetest.get_node_light(above) or 0) < 13 then
+			return
+		end
+
+		-- look for likely neighbors
+		local p2 = minetest.find_node_near(pos, 1, {"default:dirt_with_grass",
+				"default:dirt_with_dry_grass", "default:dirt_with_snow"})
+		if p2 then
+			-- but the node needs to be under air in this case
+			local n2 = minetest.get_node(above)
+			if n2 and n2.name == "air" then
+				local n3 = minetest.get_node(p2)
+				if n3 then
```

Comment:
Is this check necessary? since 'if p2 then' ensures one of those 3 nodes has been found.